sound/soc/sof/control.c

```diff
@@ -431,24 +440,38 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 		return -EINVAL;
 	}
 
-	if (copy_from_user(cdata->data, tlvd->tlv, header.length))
+	/* Allocate temporary buffer for new cdata */
+	new_cdata = kzalloc(sizeof(*new_cdata) + header.length, GFP_KERNEL);
+	if (!new_cdata)
+		return -ENOMEM;
+
+	if (copy_from_user(new_cdata->data, tlvd->tlv, header.length)) {
+		kfree(new_cdata);
 		return -EFAULT;
+	}
 
-	if (cdata->data->magic != SOF_ABI_MAGIC) {
+	if (new_cdata->data->magic != SOF_ABI_MAGIC) {
 		dev_err_ratelimited(sdev->dev,
 				    "error: Wrong ABI magic 0x%08x.\n",
-				    cdata->data->magic);
+				    new_cdata->data->magic);
+		kfree(new_cdata);
```

Comment:
There starts to be so many many exit points, that adding a label at end of function and goto to error starts to be more readable. But I'm ok with this as well.

Reply:
I don't myself like "goto". As long as OK with reviewers I'd prefer this way (just one line exit task).